src/Modal/index.jsx

```diff
@@ -0,0 +1,133 @@
+    super(props);
+
+    this.close = this.close.bind(this);
+    this.handleKeyDown = this.handleKeyDown.bind(this);
+
+    this.state = {
+      open: props.open,
+    };
+  }
+
+  componentDidMount() {
+    if (this.xButton) {
+      this.xButton.focus();
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.open !== this.props.open) {
+      this.setState({ open: nextProps.open });
+    }
+  }
+
+  componentDidUpdate(prevState) {
+    if (this.state.open && !prevState.open) {
+      this.xButton.focus();
+    }
+  }
+
+  close() {
+    this.setState({ open: false });
+    this.props.onClose();
+  }
+
+  handleKeyDown(e) {
+    if (e.key === 'Escape') {
+      this.close();
+    } else if (e.key === 'Tab') {
+      if (e.shiftKey) {
+        if (e.target === this.xButton) {
+          e.preventDefault();
+          this.closeButton.focus();
+        }
+      } else if (e.target === this.closeButton) {
+        e.preventDefault();
+        this.xButton.focus();
+      }
+    }
+  }
+
+  renderButtons() {
+    return this.props.buttons.map((button, i) => {
+      let buttonElement = button;
+      let buttonProps = button.props;
+
+      if (button.type !== Button) {
+        buttonProps = button;
+      }
+
+      buttonElement = (<Button
+        {...buttonProps}
+        key={i}
+        onKeyDown={this.handleKeyDown}
+      />);
+
+      return buttonElement;
+    });
+  }
+
+  render() {
+    return (
+      <div className={classNames(styles.modal, styles.show, { [styles['modal-open']]: this.state.open })} role="dialog">
+        <div className={styles['modal-dialog']} role="document">
+          <div className={styles['modal-content']}>
+            <div className={styles['modal-header']}>
+              <h5 className={styles['modal-title']}>{this.props.title}</h5>
+              <Button
+                display="&times;"
+                aria-label="Close"
+                buttonType="light"
+                onClick={this.close}
+                inputRef={(input) => { this.xButton = input; }}
+                onKeyDown={this.handleKeyDown}
+              />
+            </div>
+            <div className={styles['modal-body']}>
+              <p>{this.props.body}</p>
+            </div>
+            <div className={styles['modal-footer']}>
+              {this.renderButtons()}
+              <Button
+                display="Close"
+                buttonType="secondary"
+                onClick={this.close}
+                inputRef={(input) => { this.closeButton = input; }}
+                onKeyDown={this.handleKeyDown}
+              />
+            </div>
+          </div>
+        </div>
+      </div>
+    );
+  }
+}
+
+Modal.propTypes = {
+  open: PropTypes.bool,
+  title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
+  body: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
+  buttons: PropTypes.arrayOf(PropTypes.oneOfType([
+    PropTypes.element,
```

Comment:
TODO: This could use a custom validator.

Reply:
I'm kinda split on this. On the one hand, there's out-of-the-box validators we could use to make this easy (https://www.npmjs.com/package/react-element-proptypes), but on the other hand HRH Abramov [discourages typechecking that's this granular](https://github.com/facebook/react/pull/7119#issuecomment-228615386). I'm inclined to say just leave it as-is -- it'll make things easier if we end up adopting something like [Flow](https://flow.org/) instead of propTypes for typechecking.

Reply:
Sounds good to me!